tools/2023-11-interactive-commands-functional-tests.md

```diff
@@ -0,0 +1,124 @@
+# Title
+
+* **Status**: Pending
+* **Author**: Yetkin Timocin (@ytimocin)
+
+## Overview
+
+Radius CLI has multiple interactive commands that provide very important value to its users. Because these commands are very important for the functionality of Radius, we must make sure that we are able to cover different scenarios of these commands end to end. This design document will be about adding a tool to add functional tests for interactive Radius commands.
+
+## Terms and definitions
+
+Interactive Command: A command that needs to get user input. For example `rad init`.
+Pseudo-terminal: A pseudo-terminal is a software emulation of physical terminal that can be used to control terminal applications programmatically. We are going to be using pty and vt10x.
+
+## Objectives
+
+> **Issue Reference:** <https://github.com/radius-project/radius/issues/6572>
+
+### Goals
+
+* Add a tooling so that adding functional/e2e tests for interactive Radius commands is very easy for all the contributors.
+* Add a tooling so that adding functional/e2e tests for different paths of a given interactive Radius command is very easy for all the contributors.
+* Add functional/e2e tests for the interactive commands so that we know if something is broken.
+
+### Non goals
+
+### User scenarios (optional)
+
+#### User story 1
+
+We have had two problems with the installation of dev recipes during the `rad init` command run. Right after releasing Radius, we noticed that dev recipes were not being installed by the `rad init` command. This could have been easily caught if we had had a functional test for the `rad init` command in place.
+
+#### User story 2
+
+A Radius contributor would like to add a new command that requires user input which means that the command is interactive. The contributor finishes writing the code and would like to add an e2e test. At that time, the contributor can use the new interactive command test tool to write an e2e test.
+
+In the future, another contributor would like to change the flow of the interactive command that was mentioned above. The change breaks one of the e2e tests which is very good because it is caught at the pull request level.
+
+## Design
+
+* <https://github.com/Netflix/go-expect>: "An expect-like golang library to automate control of terminal or console based programs."
+* <https://github.com/hinshun/vt10x>: "About Package vt10x is a vt10x terminal emulation backend."
+* <https://github.com/creack/pty>: "PTY (PseudoTeletYpe) interface for Go."
+
+### Design details
```

Comment:
It will be helpful to provide a high level overview of where the new data types/functionality (listed in the api design section) will be added, how they will be consumed by the tests etc.